clippy_lints/src/usize_unportable_32_bit_literal.rs

```diff
@@ -0,0 +1,163 @@
+}
+
+declare_lint_pass!(UsizeUnportable32BitLiteral => [USIZE_UNPORTABLE_32_BIT_LITERAL]);
+
+/// Checks if the type is usize
+fn is_usize<'tcx>(stmt: &'tcx Stmt<'tcx>) -> bool {
+    if let StmtKind::Local(l) = stmt.kind
+        && let Some(ty) = l.ty
+    {
+        match ty.kind {
+            TyKind::Path(QPath::Resolved(_, path)) => matches!(path.res, rustc_hir::def::Res::PrimTy(Uint(Usize))),
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+/// checks if the statement contains a literal that is larger than
+/// `u32::MAX`
+fn is_larger_than_32bit<'tcx>(stmt: &'tcx Stmt<'tcx>) -> bool {
+    if let StmtKind::Local(l) = stmt.kind
+        && let Some(init) = l.init
+    {
+        match init.kind {
+            ExprKind::Lit(lit) => match lit.node {
+                LitKind::Int(v, _) => v.0 > u32::MAX.into(),
+                _ => false,
+            },
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+fn is_larger_than_32bit_literal(expr: &Expr<'_>) -> bool {
+    if let ExprKind::Lit(lit) = expr.kind {
+        match lit.node {
+            LitKind::Int(v, LitIntType::Unsigned(ty)) => v.0 > u32::MAX.into() && ty == UintTy::Usize,
+            _ => false,
+        }
+    } else {
+        false
+    }
+}
+
+/// Checks if the attribute contains `target_pointer_width` = "64"
+fn has_pointer_width_64_attr(attrs: &[Attribute]) -> bool {
+    let ident = Ident(target_pointer_width, false);
+    let str64 = TokenKind::Literal(rustc_ast::token::Lit {
+        kind: rustc_ast::token::LitKind::Str,
+        symbol: Symbol::intern("64"),
+        suffix: None,
+    });
+    let mut state = 0;
+    for attr in attrs {
+        if let Normal(normal_attr) = &attr.kind
+            && let AttrArgs::Delimited(args) = &normal_attr.item.args
+        {
+            for t in args.tokens.trees() {
+                if let Token(token, _) = t {
+                    if state == 0 && token.kind == ident {
+                        state = 1;
+                    } else if state == 1 && token.kind == TokenKind::Eq {
+                        state = 2;
+                    } else if state == 2 && token.kind == str64 {
+                        return true;
+                    } else {
+                        state = 0;
+                    }
+                }
+            }
+        }
+    }
+    false
+}
+
+/// Checks if we are in the context of a `target_pointer_width` = "64" config
+fn any_parent_has_64_pointer_width(tcx: TyCtxt<'_>, node: HirId) -> bool {
+    let map = &tcx.hir();
+    let mut prev_enclosing_node = None;
+    let mut enclosing_node = node;
+    while Some(enclosing_node) != prev_enclosing_node {
+        if has_pointer_width_64_attr(map.attrs(enclosing_node)) {
+            return true;
+        }
+        prev_enclosing_node = Some(enclosing_node);
+        enclosing_node = map.get_parent_item(enclosing_node).into();
+    }
+
+    false
+}
```

Comment:
This should use `map.parent_iter().any(...)`. Statements can have attributes, so you can't just use the parent item. Expression cannot have attributes, so you can skip the initial node (the function name also implies you do this).